src/mad_spark_alt/evolution/semantic_operators.py

```
@@ -21,62 +21,133 @@
+        logger.debug(f"Cached {operation_type} result for hash {key[:8]}")
+        
+        # Periodic cache cleanup to prevent memory growth
+        if len(self._cache) > 500:  # Limit cache size
+            self._cleanup_expired_entries()
+    
+    def _cleanup_expired_entries(self) -> None:
+        """Clean up expired cache entries to manage memory usage."""
+        current_time = time.time()
+        expired_keys = []
+        
+        for key, (value, timestamp) in self._cache.items():
```

Ruff (0.12.2)

Loop control variable `value` not used within loop body

Rename unused `value` to `_value`

(B007)